### PR TITLE
feat(rulesets): support 2.1.0, 2.2.0, 2.3.0 AsyncAPI versions

### LINKED ISSUE #1759: Update to AsyncAPI Spec 2.1.0 & 2.2.0 & 2.3.0
**User story.**
As a API designer, I can run spectral to check my AsyncAPI, nowadays I'm only able to check specs with version 2.0.0. 

**Is your feature request related to a problem?**
The new release 2.1.0 of the spec is not supported by Spectral.

**Describe the solution you'd like**
The ruleset should be extended.

https://www.asyncapi.com/blog/release-notes-2.1.0
https://www.asyncapi.com/blog/release-notes-2.2.0
https://www.asyncapi.com/blog/release-notes-2.3.0

### PR DESCRIPTION
**Pull request by magicmatatjahu**
Fixes #1759.

**Checklist**

- [X] Tests added / updated
- [ ] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

There're not a breaking changes but are some backward compatible named exports.

**Additional context**

This is my first contribution to this project (btw. awesome project!), so if I forgot something please let me know. I don't know if I should describe somewhere in the documentation the support for the latest versions of AsyncAPI, if so please let me know where.

Changes:
- remove JSON Schema for AsyncAPI 2.0.0 from source code - use `@asyncapi/specs` package where there are always up-to-date JSON Schemas.
- adding additional formats like `aas2_0`, `aas2_1` etc.
- adding/updating tests for new formats.
- adding `asyncApi2DocumentSchema` rule identical as for OpenAPI to validate the document using the appropriate JSON Schema version - I know about duplicated code so please let me know if I should create something like `common.ts` file with reusable functions for AsyncAPI and OpenAPI in the `rulesets` package.

NOTE:

I cannot run all tests locally, because I have several failed tests related to the `spectral-functions` package like:

```
  ● Linter service › when loading specification files from web › outputs warnings

    expect(received).resolves.toEqual()

    Received promise rejected instead of resolved
    Rejected to value: [Error: Cannot resolve module '@stoplight/spectral-functions' from paths ['/Users/-/Documents/Github/spectral/packages/cli/src/services/__tests__/__fixtures__', '/Users/-/Documents/Github/spectral/packages/cli/src/services/linter/utils'] from /Users/-/Documents/Github/spectral/packages/cli/src/services/__tests__/__fixtures__/spectral.js]

      299 |       });
      300 |
    > 301 |       return expect(run(`lint http://foo.local/openapi`)).resolves.toEqual([
          |              ^
      302 |         {
      303 |           code: 'info-matches-stoplight',
      304 |           message: 'Info must contain Stoplight',

      at expect (node_modules/expect/build/index.js:178:15)
      at Object.<anonymous> (packages/cli/src/services/__tests__/linter.test.ts:301:14)
``` 

Is a known issue, or perhaps it's related to my environment? 🤷🏼

### CODE REVIEW

**Comment by P0lip** [comment]
Thanks a bunch for the PR. I truly appreciate it.
I'll take a look at it today or tomorrow. Having a peek at it, it looks good to me, but I want to have a more in-depth take on it.

>  Rejected to value: [Error: Cannot resolve module '@stoplight/spectral-functions' from paths ['/Users/-/Documents/Github/spectral/packages/cli/src/services/__tests__/__fixtures__', '/Users/-/Documents/Github/spectral/packages/cli/src/services/linter/utils'] from /Users/-/Documents/Github/spectral/packages/cli/src/services/__tests__/__fixtures__/spectral.js]


Hmmm, that is odd. It did work on the CI though, so everything's good.
Perhaps try running `yarn --force` to see if that changes anything.

File: yarn.lock [hunk 1]
```diff
@@ -17,6 +17,13 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@asyncapi/specs@npm:^2.13.0":
```
Comment (P0lip) [inline yarn.lock:20]: so good to see that schema.asyncapi2.json disappear ❤️

**Comment by magicmatatjahu** [comment]
@P0lip Thanks for quick response!

> I'll take a look at it today or tomorrow. Having a peek at it, it looks good to me, but I want to have a more in-depth take on it.

No problem, no rush :) If something will be wrong (like duplicated functions) please comment and I will handle that.

> Hmmm, that is odd. It did work on the CI though, so everything's good. Perhaps try running yarn --force to see if that changes anything.

I see that on CI it pass so probably it's env problems.

**Comment by muenchhausen** [comment]
How to test that? It wasn't working for me. I tried:

A rules file:
```yml
extends: "spectral:asyncapi"
```
I just copied the sample from https://studio.asyncapi.com/:

```yml
asyncapi: '2.3.0'
info:
  title: Streetlights Kafka API
  version: '1.0.0'
  description: |
    The Smartylighting Streetlights API allows you to remotely manage the city lights.
...
```

Then I get the error:

```
  1:11    error  asyncapi-schema                 "asyncapi" property must be equal to one of the allowed values: "2.0.0"
```

A simple solution is to correct the file `packages/rulesets/src/asyncapi/schemas/schema.asyncapi2.json` and add version 2.x to be valid.

**Comment by magicmatatjahu** [comment]
@muenchhausen Hello! I tested in this way (you need to checkout to this branch):
- go to `packages/cli` folder
- create simple spectral config file like in Readme:

  ```bash
  printf '{\n  "extends": ["spectral:oas", "spectral:asyncapi"]\n}\n' > .spectral.json
  ```
  
- create AsyncAPI file (like `sample.yaml`)
- run command (in the `packages/cli` folder):

  ```yaml
  yarn cli lint {AsyncAPI file}
  ```
  
I have these errors for simple AsyncAPI file (based on `2.3.0` version):

```
/Users/---/spectral/sample.yaml
  1:1  warning  asyncapi-servers                AsyncAPI object must have non-empty "servers" object.
  1:1  warning  asyncapi-tags                   AsyncAPI object must have non-empty "tags" array.
  2:6  warning  asyncapi-info-contact           Info object must have "contact" object.                        info
  2:6  warning  asyncapi-info-license           Info object must have "license" object.                        info
 8:15  warning  asyncapi-operation-description  Operation "description" must be present and non-empty string.  channels.user/signedup.subscribe
 8:15    error  asyncapi-operation-operationId  Operation must have "operationId".                             channels.user/signedup.subscribe
```

I also tested this https://raw.githubusercontent.com/asyncapi/spec/master/examples/streetlights-kafka.yml and it also works :)

**Review by P0lip — APPROVED** [review]
Looks great! Thanks a lot.
I'll merge it tomorrow, there's one change I'll need to introduce to make sure the release goes smoothly, but I'll handle that.

File: packages/rulesets/src/asyncapi/functions/asyncApi2DocumentSchema.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  for (let i = 0; i < errors.length; i++) {
+    const error = errors[i];
+
+    if (i + 1 < errors.length && errors[i + 1].instancePath === error.instancePath) {
+      errors.splice(i + 1, 1);
+      i--;
+    } else if (i > 0 && shouldIgnoreError(error) && errors[i - 1].instancePath.startsWith(error.instancePath)) {
+      errors.splice(i, 1);
+      i--;
+    }
```
Comment (P0lip) [inline packages/rulesets/src/asyncapi/functions/asyncApi2DocumentSchema.ts:55]: most of that stuff could be indeed shared with the OAS ruleset, but we can do it later.
It's not a big deal for now.

**Comment by magicmatatjahu** [comment]
@P0lip Thanks a lot! If there will be problems/bugs, please ping me in this thread :)

We at AsyncAPI are thinking about moving (if you don't mind) all the rules related to AsyncAPI to our organization on github. That way, people could use the up-to-date ruleset for AsyncAPI and you wouldn't have to worry about maintaning AsyncAPI on your side. Here is the issue I created https://github.com/asyncapi/parser-js/issues/477. Of course I don't give a word that we will do such a thing, we have to discuss it at our place first, but if we decided to do such a thing, would Stoplight have something against it?

By the way, I'm currently testing a custom rules for more complex validations like binding validations, extensions etc and I'm delighted. Super project!

**Comment by P0lip** [comment]
@magicmatatjahu I'll let you know soon. I reached out to folks internally to see what their take on it is and will try to provide an update as soon as I learn anything.

> By the way, I'm currently testing a custom rules for more complex validations like binding validations, extensions etc and I'm delighted. Super project!

Thanks! ❤️ We're really glad you like the project.

**Pull request merged**